protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,521 @@
+            rewardData.rewardRate = uint216((amount + leftover) / rewardDuration);
+            // collect dust
+            nextRewardAmount = (amount + leftover) - (rewardData.rewardRate * rewardDuration);
+        }
+        rewardData.lastUpdateTime = uint40(block.timestamp);
+        rewardData.periodFinish = uint40(block.timestamp + rewardDuration);
+    }
+
+    function _lastTimeRewardApplicable(uint256 _finishTime) internal view returns (uint256) {
+        if (_finishTime < block.timestamp) {
+            return _finishTime;
+        }
+        return block.timestamp;
+    }
+
+    function _distributeGold() internal {
+        /// @dev no op silent fail if nothing to distribute
+        ITempleGold(address(rewardToken)).mint();
+    }
+
+    function _delegate(address delegator, address delegatee) internal {
+        address currentDelegate = delegates[delegator];
+        uint256 delegatorBalance = _balances[delegator];
+        delegates[delegator] = delegatee;
+
+        emit DelegateChanged(delegator, currentDelegate, delegatee);
+        _moveDelegates(currentDelegate, delegatee, delegatorBalance);
+    }
+
+    function _moveDelegates(
+        address srcRep,
+        address dstRep,
+        uint256 amount
+    ) internal {
+        if (srcRep != dstRep && amount > 0) {
+            if (srcRep != address(0)) {
+                uint256 srcRepNum = numCheckpoints[srcRep];
+                uint256 srcRepOld = srcRepNum > 0 ? _checkpoints[srcRep][srcRepNum - 1].votes : 0;
+                uint256 srcRepNew = srcRepOld - amount;
+                _writeCheckpoint(srcRep, srcRepNum, srcRepOld, srcRepNew);
+            }
+
+            if (dstRep != address(0)) {
+                uint256 dstRepNum = numCheckpoints[dstRep];
+                uint256 dstRepOld = dstRepNum > 0 ? _checkpoints[dstRep][dstRepNum - 1].votes : 0;
+                uint256 dstRepNew = dstRepOld + amount;
+                _writeCheckpoint(dstRep, dstRepNum, dstRepOld, dstRepNew);
+            }
+        }
+    }
+
+    function _writeCheckpoint(
+        address delegatee,
+        uint256 nCheckpoints,
+        uint256 oldVotes,
+        uint256 newVotes
+    ) internal {
+        if (nCheckpoints > 0 && _checkpoints[delegatee][nCheckpoints - 1].fromBlock == block.number) {
+            _checkpoints[delegatee][nCheckpoints - 1].votes = newVotes;
+        } else {
+            _checkpoints[delegatee][nCheckpoints] = Checkpoint(block.number, newVotes);
+            numCheckpoints[delegatee] = nCheckpoints + 1;
+        } 
+        emit DelegateVotesChanged(delegatee, oldVotes, newVotes);
+    }
+
+    modifier updateReward(address _account) {
+        {
+            rewardData.rewardPerTokenStored = uint216(_rewardPerToken());
+            rewardData.lastUpdateTime = uint40(_lastTimeRewardApplicable(rewardData.periodFinish));
+            if (_account != address(0)) {
+                claimableRewards[_account] = _earned(_account);
```

Comment:
The internal call to `_rewardPerToken()` reads multiple variables from storage. This function is called twice inside the `updateReards()` modifier, but the output of both cases will be the same as they are run in the same transaction.  (The first time is called directly in this modifier to update `rewardData.rewardsPerTokenStored`, and the second time is inside `_earned()`). 

A significant amount of gas can be saved by caching the output of `_rewardPerToken()` and passing it as an argument to `_earned()`. 

Given that this function will probably be one that will be executed more times over the lifetime of the contract I definitely think it is worth it. 

The change would look something like this:
```diff
-   rewardData.rewardPerTokenStored = uint216(_rewardPerToken());
+   uint216 rewardPerTokenCached = uint216(_rewardPerToken());
+   rewarddata.rewardPerTokenStored = rewardPerTokenCached;
// ...
-   claimableRewards[_account] = _earned(_account);
+   claimableRewards[_account] = _earned(_account, rewardPerTokenCached);  // save a lot of gas here
-   userRewardPerTokenPaid[_account] = uint256(rewardData.rewardPerTokenStored);
+   userRewardPerTokenPaid[_account] = uint256(rewardPerTokenCached); // you also save a storage-read here
```

And of course you have to redefine the `_earned()` function to accept the new argument:

```diff
    function _earned(
        address _account,
+       uint256 rewardPerToken
    ) internal view returns (uint256) {
-       return _balances[_account] * (_rewardPerToken() - userRewardPerTokenPaid[_account]) / 1e18 
            + claimableRewards[_account];
+       return _balances[_account] * (rewardPerToken - userRewardPerTokenPaid[_account]) / 1e18 
            + claimableRewards[_account];
    }
```

Reply:
Good!